llvm-external-projects/iree-dialects/lib/Dialect/VectorExt/IR/VectorExtOps.cpp

```diff
@@ -52,7 +52,84 @@ OpFoldResult ToSIMTOp::fold(FoldAdaptor) {
   return {};
 }
 
+LayoutIterator &LayoutIterator::operator++() {
+  for (auto &[dim, it] : state) {
+    if (frozenDimensions.contains(dim))
+      continue;
+    if (it != ranges[dim].end()) {
+      ++it;
+      break;
+    }
+    it = ranges[dim].begin();
+  }
+  return *this;
+}
+
+void LayoutIterator::maybeFreezeAndConcatenate(
+    const LayoutIterator &frozenIterator) {
+  for (auto &[frozenDim, frozenIt] : frozenIterator.getState()) {
+    if (!state.contains(frozenDim)) {
+      frozenDimensions.insert(frozenDim);
+      state[frozenDim] = frozenIt;
+    }
+  }
+}
+
+static bool isLaneDimension(LayoutDimension dim) {
+  return (dim == LayoutDimension::LANEX) || (dim == LayoutDimension::LANEY) ||
+         (dim == LayoutDimension::LANEZ);
+}
+
+void LayoutIterator::initialize(PerDimLayoutAttr &attr,
+                                DenseMap<LayoutDimension, int64_t> strides) {
+  auto reversedLabels = llvm::reverse(attr.getLabels());
+  auto reversedShapes = llvm::reverse(attr.getShapes());
+  for (auto [nameAttr, shape] : llvm::zip(reversedLabels, reversedShapes)) {
+    LayoutDimension dim = nameAttr.getValue();
+    if (isLaneDimension(dim))
+      continue;
+    int64_t stride = strides.contains(dim) ? strides[dim] : 1;
+    ranges[dim] = DimensionalRange(0, shape - 1, stride);
+    state[dim] = ranges[dim].begin();
+  }
+}
+
+LayoutIterator::LayoutIterator(LayoutAttr &attr,
+                               DenseMap<LayoutDimension, int64_t> strides) {
+  for (PerDimLayoutAttr perDimAttr : attr.getLayouts()) {
+    initialize(perDimAttr, strides);
+  }
+}
+
+LayoutIterator::LayoutIterator(PerDimLayoutAttr &attr,
+                               DenseMap<LayoutDimension, int64_t> strides) {
+  initialize(attr, strides);
+}
+
+/// The iterator is done when it returns back to
+/// its begin state.
+bool LayoutIterator::iterationComplete() {
```

Comment:
nit: iterationComplete -> isIterationComplete

Reply:
So I was thinking about how this reads
```
do {
  ...
}while (!iterationComplete());
```
which is _do something while not iteration complete_ which to me reads better than _do something while not is iteration complete_.
```suggestion
bool LayoutIterator::iterationComplete() {
```
But can change if you feel strongly.